### DISCUSSION TITLE
CodeQl is not identifying password in plain text in c# code

### DISCUSSION THREAD

**Discussion by umaranit** [community/community#21441] [Other Feature Feedback, Questions, & Ideas]
<p>I have a sample c# application which has password in plain text. I have configured a codeql workflow. The workflow only identifies few javascript issues but its not identifying the plain password security issue present in c# application</p>

**Comment by criemen** [comment]
<p>Hi,<br>
we have the <code>cs/hardcoded-credentials</code> query which should catch this.<br>
If the code is not proprietary, would you mind sharing the sample project and the workflow you used to analyze the project with codescanning?</p>

**Comment by umaranit** [comment]
<p>Hi Please find below the repo details</p>
<aside class="onebox allowlistedgeneric" data-onebox-src="https://github.com/umaranit/corewebapp">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">

      <a href="https://github.com/umaranit/corewebapp" target="_blank" rel="noopener nofollow ugc">GitHub</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/345;"><img src="https://user-images.githubusercontent.com/22909641/181064555-2222afd2-ab1e-40a9-93af-94f89a305083.png" class="thumbnail" width="690" height="345" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/2/8/28338e36e82d811b01d6e082549ade0edbdfb7bc_2_690x345.png, https://aws1.discourse-cdn.com/github/optimized/3X/2/8/28338e36e82d811b01d6e082549ade0edbdfb7bc_2_1035x517.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/2/8/28338e36e82d811b01d6e082549ade0edbdfb7bc.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/2/8/28338e36e82d811b01d6e082549ade0edbdfb7bc_2_10x10.png"></div>

<h3><a href="https://github.com/umaranit/corewebapp" target="_blank" rel="noopener nofollow ugc">GitHub - umaranit/corewebapp</a></h3>

  <p>Contribute to umaranit/corewebapp development by creating an account on GitHub.</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>Refer the codeql-analysis.yml in it</p>

**Comment by kingthorin** [comment]
<p>Based on: <a href="https://github.com/umaranit/corewebapp/runs/3430720607?check_suite_focus=true" class="inline-onebox" rel="noopener nofollow ugc">updated con string · umaranit/corewebapp@88b06d6 · GitHub</a></p>
<pre><code class="lang-auto">Compiling query plan for /opt/hostedtoolcache/CodeQL/0.0.0-20210809/x64/codeql/qlpacks/codeql-csharp/Security Features/CWE-798/HardcodedCredentials.ql.
Resolving imports for /opt/hostedtoolcache/CodeQL/0.0.0-20210809/x64/codeql/qlpacks/codeql-csharp/Security Features/CWE-798/HardcodedCredentials.ql.
Compilation cache hit for /opt/hostedtoolcache/CodeQL/0.0.0-20210809/x64/codeql/qlpacks/codeql-csharp/Security Features/CWE-798/HardcodedCredentials.ql.
[25/166] Found in cache: /opt/hostedtoolcache/CodeQL/0.0.0-20210809/x64/codeql/qlpacks/codeql-csharp/Security Features/CWE-798/HardcodedCredentials.ql.

Starting evaluation of codeql-csharp/Security Features/CWE-798/HardcodedCredentials.ql.

[56/166 eval 58.1s] Evaluation done; writing results to codeql-csharp/Security Features/CWE-798/HardcodedCredentials.bqrs.
[57/166 eval 1m2s] Evaluation done; writing results to codeql-csharp/Security Features/CWE-798/HardcodedConnectionString.bqrs.

Intepreted pathproblem query "Hard-coded connection string with credentials" (cs/hardcoded-connection-string-credentials) at path /opt/hostedtoolcache/CodeQL/0.0.0-20210809/x64/codeql/qlpacks/codeql-csharp/Security Features/CWE-798/HardcodedConnectionString.ql.
Interpreting /opt/hostedtoolcache/CodeQL/0.0.0-20210809/x64/codeql/qlpacks/codeql-csharp/Security Features/CWE-798/HardcodedCredentials.ql...
... found results file at /home/runner/work/_temp/codeql_databases/csharp/results/codeql-csharp/Security Features/CWE-798/HardcodedCredentials.bqrs.
Intepreted pathproblem query "Hard-coded credentials" (cs/hardcoded-credentials) at path /opt/hostedtoolcache/CodeQL/0.0.0-20210809/x64/codeql/qlpacks/codeql-csharp/Security Features/CWE-798/HardcodedCredentials.ql.
</code></pre>
<p>I only noticed the hard coded connect string item part way through my search, but anyway “pathproblem” might give you a hint about what’s going on, or something to search :man_shrugging: Having the q1 and bqrs references might give you a lead to lookup the actual queries and see why your item isn’t being picked up.</p>

**Comment by umaranit** [comment]
<p>Hi</p>
<p>I am not clear. Have I done any wrong configuration in the workflow? Which area do I need to investigate?</p>

**Comment by criemen** [comment]
<p>Hi,<br>
so I had a look at your source code.<br>
Our queries try to detect real-world vulnerabilities. Your project currently is not vulnerable because you never really use the hardcoded password (the only thing you have is a dead read).<br>
This means that you have to use the hardcoded password somewhere for us to detect it.</p>
<p>In CodeQL, we have a concept of dataflow, which describes where data comes from (a source, in your case a string literal), and where it flows to (a sink).<br>
In your case, you are missing a sink.<br>
Sinks in the <code>HardcodedCredentials</code> query are for example library methods (i.e. coming from a dotnet assembly, not your own project) with parameters that have the name password or passphrase.<br>
I hope that helps!</p>

**Comment by kingthorin** [comment]
<p><a class="mention" href="/criemen">@criemen</a> whats with the “pathproblem “ string in the logs? Is that just an unfortunate set of words for people unfamiliar with CodeQL?</p>

**Comment by umaranit** [comment]
<p>ok. Got it. Thanks for the details</p>

**Comment by criemen** [comment, marked as answer]
<p><a class="mention" href="/kingthorin">@kingthorin</a> Ah, I think you got side-tracked here.<br>
This <code>pathproblem</code> string does not refer to a problem during query evaluation at all, it rather refers to the query <code>@kind</code> attribute. See the documentation <a href="https://codeql.github.com/docs/writing-codeql-queries/about-codeql-queries/#query-metadata">here</a> if you’re interested to learn more about query kinds.<br>
pathproblem means that the query looks for problems in your code that are associated with one or more dataflow paths.</p>
<p>I hope that helps!</p>

**Comment by kingthorin** [comment]
<p>Thank you, that does help.</p>
